### ISSUE TITLE
Migrate project system IVTs to external access pattern

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#35070]

**Comment by sharwell** [comment]
This is a complete list of internals currently used in dotnet/project-system:

* Microsoft.Internal.Performance
  * CodeMarkers
* Microsoft.VisualStudio.LanguageServices.Implementation.EditAndContinue.Interop
  * ENC_BREAKSTATE_REASON
  * IVsENCRebuildableProjectCfg2
  * IVsENCRebuildableProjectCfg4
* Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
  * IEntryPointFinderService
* Microsoft.VisualStudio.LanguageServices.ProjectSystem
  * ICodeModelFactory
  * ITempPECompiler
  * IWorkspaceProjectContext
  * IWorkspaceProjectContextFactory

**Comment by Pilchie** [comment]
I propose that we just make the types required above public and remove the IVT entirely instead.

**Comment by sharwell** [comment]
As for the use of `ProjectSystemContractAttribute`, I believe these attributes can be updated to reference the MEF wrapper type instead of the original, should we choose to not make the types public.

> I propose that we just make the types required above public and remove the IVT entirely instead.

I have prepared the external access layer, which _allows_ us to consider this proposal separately from the IVT elimination process, in the event it's not obvious that these types are ready to be stable APIs.

**Comment by tmat** [comment]
I don't think we should make any of these types public. We will need to significantly modify them for OOP work and making them public afterwards would be a breaking change.

**Comment by tmat** [comment]
As a general rule we don't want any new public types in Roslyn VS layer.

**Comment by davkean** [comment]
IWorkspaceProjectContext is an exchange type and is exchanged with other Roslyn APIs and cast to a variety of interfaces. Please factor that in your design.

**Comment by sharwell** [comment]
@davkean We'll have to come up with a new way to handle that specific case. I'm not sure what it will look like yet.
